linkerd/trace-context/src/propagation/w3c.rs

```diff
@@ -0,0 +1,148 @@
+use http::HeaderValue;
+use tracing::{trace, warn};
+
+use super::{decode_id_with_padding, get_header_str, Propagation, TraceContext};
+use crate::{Flags, Id};
+
+pub const HTTP_TRACEPARENT: &str = "traceparent";
+pub const HEADER_VERSION_00: &str = "00";
+
+pub fn unpack_w3c_trace_context<B>(request: &http::Request<B>) -> Option<TraceContext> {
+    get_header_str(request, HTTP_TRACEPARENT).and_then(parse_context)
+}
+
+/// Given an http request and a w3c trace context, create a new Span ID and
+/// assign it to the tracecontext header value, in order to propagate
+/// the trace context downstream.
+pub fn increment_http_span_id<B>(request: &mut http::Request<B>, context: &TraceContext) -> Id {
+    let span_id = Id::new_span_id(&mut rand::thread_rng());
+
+    trace!("incremented span id: {}", span_id);
+
+    let new_header = {
+        let mut buf = String::with_capacity(60);
+        buf.push_str(HEADER_VERSION_00);
+        buf.push('-');
+        buf.push_str(&hex::encode(context.trace_id.as_ref()));
+        buf.push('-');
+        buf.push_str(&hex::encode(context.parent_id.as_ref()));
+        buf.push('-');
+        buf.push_str(&hex::encode(vec![context.flags.0]));
```

Comment:
ah, looks like there's already a PR for that upstream (https://github.com/KokaKiwi/rust-hex/pull/66) but the maintainer of `hex` seems to have largely abandoned the crate. it's _probably_ not worth a fork just to land that change (and perhaps also https://github.com/KokaKiwi/rust-hex/pull/64), but...

Reply:
We could encode into a bytes buffer maybe? https://docs.rs/hex/latest/hex/fn.encode_to_slice.html. But I think the logic would be a bit hairy. What I have in mind is something like:

```rust
// Sz depends on id lengths * 2 (to allow for hex encoding) + 2 u8s for version, + 2 u8s for flags and 3 separators
let buf_sz = (trace_id.len() + parent_id.len()) * 2 + 7;
let mut buf = Vec::with_capacity(buf_sz);
buf.push(0);
buf.push(0);
buf.push('-' as u8);
let mut offset = 3;
hex::encode_to_slice(trace_id, &mut buf[offset..trace_id.len()]);
```

We could technically avoid allocating any additional Strings. However, at least in the brute way that I thought of it, the logic is a bit more complicated and would likely have to be commented out. It will also be very specific, it will only apply to writing w3c contexts. Additionally, we need to keep an offset so we can slice the buffer up.

Alternatively, we could put everything into a vec and call `ToHex::encode` on it. We can still use the trait for a bit more freedom (and I think they provide blanket impls for `impl AsRef<u8>`). The problem is that would also encode our separators and we don't want that.

If we think we want to do less allocations I'm happy to explore this a bit further. Wonder if we can use the `Bytes` crate to simplify everything?

Reply:
my preference would probably be to move forwards with this PR as-is, and if you want to spend additional time on this, open a follow-up PR to change this to use a single buffer that everything is written into, the way you described. 

if we do that, we would probably want to consider adding tests that check that the encoded output matches the naive implementation that uses `hex::encode` rather than manually slicing up the buffer.